migration/src/main/scala/migration/Migrator.scala

```diff
@@ -0,0 +1,47 @@
+package migration
+
+import migration.data._
+import migration.dsl.interpreters.PgTable
+import migration.interop.MonadIO
+import migration.interop.MonadIO._
+import org.mongodb.scala._
+import scalaz._, Scalaz._
+import scalaz.zio.{ IO, App }
+import scalaz.zio.interop.scalaz72._
+import scalaz.zio.interop.console.scalaz._
+
+trait Migrator {
+
+  def WINDOW_SIZE: Int
+  
+  def migrate[A](
+    cfg: Config, 
+    mongoDb: MongoDatabase, 
+    mongoTable: String, 
+    pgTable: PgTable[A]
+  )(implicit F: FromMongo[A], T: ToPostgre[A]): Migration[MigrationError, Unit] =
+    for {
+      _ <- putStr(s"Starting migration $mongoTable -> ${pgTable.name}... ").leftMap[MigrationError](ConsoleError(_)).lift[Migration]
+      coll <- getCollection(mongoDb, mongoTable)
+      n <- getCount(coll)
+      _ <- putStrLn(s"Found $n items to migrate").leftMap[MigrationError](ConsoleError(_)).lift[Migration]
+      _ <- dropTable(pgTable.name)
+      _ <- createTable(pgTable.name, pgTable.id, pgTable.idType)
+
+      iterations = Math.ceil(n / WINDOW_SIZE).toInt
+
+      _ <- (0 until iterations).toList.traverse { i =>
+        val skip = i * WINDOW_SIZE
+        val limit = WINDOW_SIZE.toLong.min(n - skip.toLong).toInt
+        for {
+          _ <- putStr(s"Moving items ${skip} to ${skip + limit}...").leftMap[MigrationError](ConsoleError(_)).lift[Migration]
+          // TODO log errors
+          recs <- getCursor[A](coll, skip, limit)
```

Comment:
If I understand correctly the mongo collection is "sliced" into portions of size "WINDOW_SIZE".
The conditions for this to work are:
 - Mongo guarantees the order of its collection
 - Absolutely no write operation can be done on the collection during the migration as this might affect the order, and lead to data loss

I don't know if the first one is true, and the second can be enforced by turning the UI off during the migration, but wouldn't there be a way to use the same cursor for the whole table? That may or may not be desirable so I'm more asking the question rather than suggesting a solution.

Reply:
🤦‍♂️  great observation! I'll update and move the slicing _after_ the point of backpressure.